json/test_yajl.c

```diff
@@ -0,0 +1,211 @@
+#include <libnotify.h>
+#include <stdbool.h>
+#include <stdio.h>
+#include <stdlib.h>
+#include <string.h>
+
+#include <errno.h>
+#include <fcntl.h>
+#include <sys/stat.h>
+#include <sys/types.h>
+#include <unistd.h>
+
+#include <yajl/yajl_parse.h>
+#include <yajl/yajl_version.h>
+
+#if (YAJL_MAJOR != 2)
+#error "yajl 2.x is required"
+#endif
+
+#ifdef __clang__
+#define COMPILER "clang"
+#else
+#define COMPILER "gcc"
+#endif
+
+
+struct coordinate_s {
+	double x;
+	double y;
+	double z;
+};
+
+typedef struct coordinate_s coordinate_t;
+
+struct context_s {
+	size_t depth;
+	size_t len;
+	char item;
+	bool in_coord;
+
+	coordinate_t value;
+};
+
+typedef struct context_s context_t;
+
+int cb_double(void* ctx_v, double value) {
+	context_t* ctx = ctx_v;
+
+	switch (ctx->item) {
+		case 'x': ctx->value.x += value; break;
+		case 'y': ctx->value.y += value; break;
+		case 'z': ctx->value.z += value; break;
+		default:;
+	}
+
+	return 1;
+}
+
+int cb_start_map(void* ctx_v) {
+	context_t* ctx = ctx_v;
+
+	ctx->depth++;
+
+	if (ctx->depth == 2 && ctx->in_coord) {
+		ctx->len++;
+	}
+
+	return 1;
+}
+
+int cb_map_key(void* ctx_v, const unsigned char* key, size_t length) {
+	context_t* ctx = ctx_v;
+
+	if (ctx->depth == 1) {
+		ctx->in_coord = strncmp("coordinates", (const char*)key, length) == 0;
+	} else if (ctx->depth == 2 && ctx->in_coord) {
+		ctx->item = (length > 0 ? key[0] : '\0');
+	}
+
+	return 1;
+}
+
+int cb_end_map(void* ctx_v) {
+	context_t* ctx = ctx_v;
+
+	ctx->depth--;
+
+	return 1;
+}
+
+int cb_end_array(void* ctx_v) {
+	context_t* ctx = ctx_v;
+
+	if (ctx->depth == 1 && ctx->in_coord) {
+		ctx->value.x /= ctx->len;
+		ctx->value.y /= ctx->len;
+		ctx->value.z /= ctx->len;
+	}
+
+	return 1;
+}
+
+static yajl_callbacks ycallbacks = {
+	NULL, /* null */
+	NULL, /* boolean */
+	NULL, /* integer */
+	&cb_double,
+	NULL, /* number */
+	NULL, /* string */
+	&cb_start_map,
+	&cb_map_key,
+	&cb_end_map,
+	NULL, /* start_array */
+	&cb_end_array
+};
+
+context_t init_context() {
+	return (context_t){
+		.depth = 0,
+		.len = 0,
+		.item = '\0',
+		.in_coord = false,
+		.value = {
+			.x = 0.0,
+			.y = 0.0,
+			.z = 0.0
+		}
+	};
+}
+
+int verify(const char* buf, size_t size, const coordinate_t* expected) {
+	context_t ctx = init_context();
+
+	yajl_handle yajl = yajl_alloc(&ycallbacks, NULL, &ctx);
+
+	yajl_parse(yajl, (const unsigned char*)buf, size);
+	yajl_complete_parse(yajl);
+
+	yajl_free(yajl);
+
+	if (ctx.value.x != expected->x || ctx.value.y != expected->y || ctx.value.z != expected->z) {
+		fprintf(stderr, "{%lf, %lf, %lf} != {%lf, %lf, %lf}\n", ctx.value.x, ctx.value.y, ctx.value.z, expected->x, expected->y, expected->z);
```

Comment:
A little bit code duplication here. Other tests use their own dump function for that, I guess it's possible to use it too:

```c
void dump(char *buf, coordinate_t coord);

char str_x[50];
dump(str_x, ctx.value);

char str_y[50];
dump(str_y, expected);

fprintf(stderr, "%s != %s\n", str_x, str_y);
```